File: python3/symmetric_tree.py
"""
Space   : O(n)
Time    : O(n)
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
        if not root.left and not root.right:
            return True

        if root.left and root.right:
            stack = [(root.left, "l"), (root.right, "r")]
        else:
            return False

        while stack:
            if len(stack) % 2 == 1:
                return False

            l = stack.pop(0)
            r = stack.pop()

            if l[0].val != r[0].val or l[1] == r[1]:
                return False
            if l[0].right:
                stack.insert(0, (l[0].right, "r"))
            if l[0].left:
                stack.insert(0, (l[0].left, "l"))
            if r[0].left:
                stack.append((r[0].left, "l"))
            if r[0].right:
                stack.append((r[0].right, "r"))

        return True

File: golang/symmetric_tree.go
package main

import "golang/model"

// Space	: O(n)
// Time		: O(n)

func isSymmetricBFS(root *model.TreeNode) bool {
	stack := []*model.TreeNode{root}

	for len(stack) > 0 {
		temp := []*model.TreeNode{}
		for _, node := range stack {
			if node != nil {
				temp = append(temp, node.Left)
				temp = append(temp, node.Right)
			}
		}

		i, j := 0, len(temp)-1
		for i < j {
			if temp[i] != nil && temp[j] != nil {
				if temp[i].Val != temp[j].Val {
					return false
				}
			} else if temp[i] != nil && temp[j] == nil {
				return false
			} else if temp[i] == nil && temp[j] != nil {
				return false
			}
			i++
			j--
		}

		stack = temp
	}

	return true
}

// Space	: O(1)
// Time		: O(n)

func isSymmetricDFS(root *model.TreeNode) bool {
	return checkIfSymmetric(root.Left, root.Right)
}

func checkIfSymmetric(left *model.TreeNode, right *model.TreeNode) bool {
	if left == nil && right == nil {
		return true
	}

	if left == nil || right == nil {
		return false
	}

	if left.Val != right.Val {
		return false
	}

	return checkIfSymmetric(left.Left, right.Right) && checkIfSymmetric(left.Right, right.Left)
}
